### PR TITLE
ARROW-7021: [Java] UnionFixedSizeListWriter decimal type should check writer index

### PR DESCRIPTION
**Pull request by tianchen92**
Related to [ARROW-7021](https://issues.apache.org/jira/browse/ARROW-7021).

UnionFixedSizeListWriter should check writer index for decimal type (just as other types) to ensure the values written not exceed listSize.

Otherwise, the writer may continue to write data into it’s underlying vector quietly even the the writer.idx() > listSize * index.

### CODE REVIEW

File: java/vector/src/main/codegen/templates/UnionFixedSizeListWriter.java [hunk 1]
```diff
@@ -197,7 +206,7 @@ public void writeDecimal(BigDecimal value) {
   @Override
   public void write${name}(<#list fields as field>${field.type} ${field.name}<#if field_has_next>, </#if></#list>) {
     if (writer.idx() >= (idx() + 1) * listSize) {
-      throw new IllegalStateException(String.format("values at index %s is greater than listSize %s", idx(), listSize));
```
Comment (emkornfield) [inline java/vector/src/main/codegen/templates/UnionFixedSizeListWriter.java]: please revert this and use the same language above.

File: java/vector/src/test/java/org/apache/arrow/vector/TestFixedSizeListVector.java [hunk 2]
```diff
@@ -265,6 +267,26 @@ public void testUnionFixedSizeListWriter() throws Exception {
     }
   }
 
+  @Test
+  public void testDecimalIndexCheck() throws Exception {
+    try (final FixedSizeListVector vector1 = FixedSizeListVector.empty("vector", 3, allocator)) {
```
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/TestFixedSizeListVector.java]: please comment /* param_name= */ 3

File: java/vector/src/main/codegen/templates/UnionFixedSizeListWriter.java (hunk 1 above)
Comment (tianchen92) [reply java/vector/src/main/codegen/templates/UnionFixedSizeListWriter.java]: ok, fixed.

File: java/vector/src/test/java/org/apache/arrow/vector/TestFixedSizeListVector.java (hunk 2 above)
Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/TestFixedSizeListVector.java]: fixed.

File: java/vector/src/test/java/org/apache/arrow/vector/TestFixedSizeListVector.java [hunk 3]
```diff
@@ -265,6 +267,26 @@ public void testUnionFixedSizeListWriter() throws Exception {
     }
   }
 
+  @Test
```
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/TestFixedSizeListVector.java]: it doesn't look like there is a check to ensure writing succeeds?  could you also add one here?

Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/TestFixedSizeListVector.java]: sure, fixed.

**Pull request closed without merging**